### PR TITLE
Form 

### PR DESCRIPTION
**Pull request by Soare-Robert-Daniel**
### **Preview Parent**: 
image

### **Preview Child**:
image

## Saving the data

@HardeepAsrani, where and how should we save the data filled by the website users from the contact form? And how to display it to the owner of the site?

- [x] It can register as a Form Block with different child blocks, as Form Input, Form Textarea, and such.
- [x] Of these child blocks, we can have different variants to make things easier for the user. For example, Form Input can have different Text Input, Email Input, and other variants.
- [x] The Form Block will be a placeholder block that will have three variants: Contact, Newsletter & Feed Form. Each form will have its initial set of innerBlocks, while Form Block itself will be empty by default with a placeholder shown for the user to pick a default type of form.
- [x] By default email should be sent to site's user but it can be changed from the Form Block's inspector.
- [x] All variants should have a "Agree to terms and conditions" checkbox by default.
- [ ] Once we have contact form ready, we can try with Mailchimp, Sendinblue & Follow It to make Newsletter/Form Feed variants.

### CODE REVIEW

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel We won't be saving any data and the email should go to the user's email. By default, it should be the website email but we should give them an option to send it to a different email.

Also we don't really need to give a lot of CSS options, just some good default styles and leave rest to the theme.

**Comment by selul** [comment]
@HardeepAsrani your plan looks good for me. 
Captcha can be included in this version too, as a separate field.

**Comment by Soare-Robert-Daniel** [comment]
Test GH CLI﻿

**Comment by HardeepAsrani** [comment]
Hey @Soare-Robert-Daniel,

I made few changes here:

- Removed extra attributes from the Form block, like templateType and URL as those aren't needed. Templates type shouldn't be decided with an attribute but using Block Variation with a picker like https://github.com/WordPress/gutenberg/tree/master/packages/block-editor/src/components/block-variation-picker
- Input block now has 3-4 different types. Removed Checkbox as it will need a block of its own (we can skip it for this version), and extracted Textarea into a new block as it has a different structure than rest of input fields. Also removed password type as contact forms aren't the place for passwords.
- Variation can be registered from the index.js file straight, plugins aren't needed for it.

Let me know your thoughts. And you can continue your work with the block.

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel You should use AJAX for this intead of REST.

**Comment by HardeepAsrani** [comment]
Also, make sure always to include the page link that was used to send the email. As if the form is in "about page" of the website, it should also include the link to that page so it's easier for user to know which page is it coming from.

Also, "subject" attribute name should be good enough, what is email title? 😜

File: inc/class-main.php [hunk 1]
```diff
@@ -131,6 +138,9 @@ public function init() {
 		add_action( 'init', array( $this, 'load_server_side_blocks' ), 11 );
 		add_action( 'init', array( $this, 'register_meta' ), 11 );
 
+		add_action( 'wp_ajax_get_otter_form', 'get_otter_form' );
```
Comment (HardeepAsrani) [inline inc/class-main.php:141]: Please remove this.

File: inc/class-main.php [hunk 2]
```diff
@@ -557,7 +567,19 @@ public function enqueue_dependencies( $post = null ) {
...
 
-			self::$is_circle_counter_loaded = true;
+			self::$is_tabs_loaded = true;
+		}
+
+		if ( ! self::$is_form_loaded && has_block( 'themeisle-blocks/form', $post ) ) {
+			wp_enqueue_script(
+				'themeisle-gutenberg-tabs',
+				plugin_dir_url( $this->get_dir() ) . 'build/form.js',
+				array( 'wp-i18n', 'wp-dom-ready', 'jquery', 'wp-api' ),
```
Comment (HardeepAsrani) [inline inc/class-main.php:577]: If possible, can we get rid of jQuery?

File: inc/plugins/class-options-settings.php [hunk 3]
```diff
@@ -81,8 +81,40 @@ public function register_settings() {
...
 			)
 		);
+
+		register_setting(
+			'themeisle_blocks_settings',
+			'themeisle_blocks_form_emails',
+			array(
+				'type'              => 'array',
+				'description'       => __( 'Email used in the Form block.', 'otter-blocks' ),
+				'sanitize_callback' => '' /* function( $array ) {
```
Comment (HardeepAsrani) [inline inc/plugins/class-options-settings.php:91]: This needs to be added.

File: inc/server/class-form-server.php [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+		add_action( 'rest_api_init', array( $this, 'register_routes' ) );
+	}
+
+	/**
+	 * Register REST API route
+	 */
+	public function register_routes() {
+		$namespace = $this->namespace . $this->version;
+
+		register_rest_route(
```
Comment (HardeepAsrani) [inline inc/server/class-form-server.php:49]: Why do we have a method for GET?

File: inc/server/class-form-server.php [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+	 */
+	public function get_form_data( $request ) {
+
+		$return = array(
+			'success' => false,
+		);
+
+		$data = json_decode( $request->get_body(), true );
+
+		$email_subject = '[Otter] ' . ( isset( $data['emailSubject'] ) ? $data['emailSubject'] : __( 'A new submission', 'otter-blocks' ) );
```
Comment (HardeepAsrani) [inline inc/server/class-form-server.php:95]: Default subject can be "A new form submission on (website name)," also remove "[Otter]" prefix.

File: inc/server/class-form-server.php [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+			foreach( $form_emails as $form ) {
+				if( $form['form'] === $option_name ) {
+					$to = $form['email'];
+				}
+			}
+		}
+
+		$from    = sanitize_email( get_site_option( 'admin_email' ) );
+		$headers = array( 'Content-Type: text/html; charset=UTF-8', 'From: Admin <' . $from . '>' );
```
Comment (HardeepAsrani) [inline inc/server/class-form-server.php:113]: From website name or something better, Admin doesn't look good in email.

File: inc/server/class-form-server.php [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	/**
+	 * Body template preparation
+	 *
+	 * @param array $data Data from the forms.
+	 *
+	 * @return string
+	 */
+	private function prepare_body( $data ) {
```
Comment (HardeepAsrani) [inline inc/server/class-form-server.php:132]: We should include the page link in the body, a link like:

Form submission from: https://example.com/contact

**Review by HardeepAsrani — CHANGES_REQUESTED** [review]
I've added some line comments that you can go through.

Well done. We should also add form variation for "Contact form" and "Request a quote." We can have a newsletter in the next version as I'd like to explore some newsletter plugins. Make sure you use the __experimentalBlockVariationPicker  for the Placeholder.

File: inc/server/class-form-server.php (hunk 4 above)
Comment (Soare-Robert-Daniel) [reply inc/server/class-form-server.php:49]: It was for testing

File: inc/class-main.php (hunk 2 above)
Comment (Soare-Robert-Daniel) [reply inc/class-main.php:577]: yes

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache we need a test on this PR

This block will require a special setup since it needs to send emails.

To add the email functionality, add the following plugin: https://wordpress.org/plugins/wp-mail-smtp/

They support multiple providers (and have tutorials on them when you set up one ); I recommend the GMAIL since it is the one I tested.

The block support to send the email to another address. You can use 10minute mail or our emails from slack.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Nice work 🚀  Tested it and found a few issues:

- [ ] The form input boxes look different in editor and frontend
editor: https://vertis.d.pr/rZvOhv
frontend: https://vertis.d.pr/hh2c6D

- [ ] The input box border has different colors https://vertis.d.pr/3UpGrg
- [ ] After hitting submit button, i get this error in console https://vertis.d.pr/qUs5t5 (after about 5 seconds)
- [ ] The **Success** message is not that visible, makes it hard to know that the form was submitted. https://vertis.d.pr/ezmIAC
- [ ] The **message input** font is different than for other inputs https://vertis.d.pr/eVzGnU
- [ ] The email subject is always _**A new form submission on..**_ , even if i change it to something else in form settings

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, the style difference depends on the theme and not the block; all we do is the layout. The other depends on the theme and browser.

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache made an update for the email subject problem

**Comment by HardeepAsrani** [comment]
@irinelenache Let us know when you're done with this.

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel Remember to put the issue in QA column so QA team can easily track it.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Tested again and some issues are still present:

> The message input font is different than for other inputs https://vertis.d.pr/eVzGnU
> The input box border has different colors https://vertis.d.pr/3UpGrg

If it's ok to let them like this, we can move this to **Ready to merge** 👍

**Pull request merged**